### PR TITLE
perf(s3stream): reduce lock granularity in `StreamMetadataManager`

### PR DESCRIPTION
**Pull request by lifepuzzlefun**
1. use volatile field to store MetadataImage avoid acquire lock when kraft metadata update
2. when access metadata store MetadataImage to local field without hold locks.
3. for the index structure in MetadataImage use lock to trigger index build and after build success the access is volatile access.
4. avoid get list.size() in for iter.

##### test
1. add concurrent access metadata test to check the access is right.

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/log/stream/s3/metadata/StreamMetadataManager.java [hunk 1]
```diff
@@ -222,4 +237,28 @@ public GetObjectsTask() {
         }
     }
 
+    private static class S3StreamImagePair {
```
Comment (superhx) [inline core/src/main/scala/kafka/log/stream/s3/metadata/StreamMetadataManager.java:240]: Why do we need S3StreamImagePair? Images themselves are immutable.

File: core/src/main/scala/kafka/log/stream/s3/metadata/StreamMetadataManager.java [hunk 2]
```diff
@@ -74,118 +65,142 @@ public String name() {
...
+                                        "and search in metadataCache success with result: {}",
+                                streamId, startOffset, endOffset, limit, rst);
+                    }
+
+                    return inRangeObjects;
+                });
     }
 
-    private synchronized CompletableFuture<InRangeObjects> fetch0(long streamId, long startOffset, long endOffset, int limit) {
+    private CompletableFuture<Map.Entry<InRangeObjects, S3StreamImagePair>> fetch0(long streamId, long startOffset, long endOffset, int limit) {
```
Comment (superhx) [inline core/src/main/scala/kafka/log/stream/s3/metadata/StreamMetadataManager.java:128]: We can move the logic of filling S3Object information from #fetch to #fetch0, so we don't need to return image.

File: core/src/main/scala/kafka/log/stream/s3/metadata/StreamMetadataManager.java (hunk 1 above)
Comment (lifepuzzlefun) [reply core/src/main/scala/kafka/log/stream/s3/metadata/StreamMetadataManager.java:240]: yes those were immutable. we need atomic replace for the two stream relate metadata Images. 
this package can volatile update and read store in local variable without locks (for multi fields)

File: core/src/main/scala/kafka/log/stream/s3/metadata/StreamMetadataManager.java (hunk 2 above)
Comment (lifepuzzlefun) [reply core/src/main/scala/kafka/log/stream/s3/metadata/StreamMetadataManager.java:128]: I think the image should be in the same version when get metadata. this pr won't block image update and image retrieve logic. so when the metadata update the images are not the same version which will cause metadata inconsistency.

Comment (lifepuzzlefun) [reply core/src/main/scala/kafka/log/stream/s3/metadata/StreamMetadataManager.java:128]: yeah i understand what you said. i will check it out

Comment (lifepuzzlefun) [reply core/src/main/scala/kafka/log/stream/s3/metadata/StreamMetadataManager.java]: updated

**Comment by lifepuzzlefun** [comment]
if more unit test need to be added please let me know : - )

**Review by superhx — APPROVED** [review]

**Pull request merged**